src/engine/Match_tainting_mode.mli

```diff
@@ -30,8 +30,23 @@ val hook_setup_hook_function_taint_signature :
   *   (or we could infer a signature for them too...).
   *)
 
+module Formula_tbl : Hashtbl.S with type key = Rule.formula
+
+val mk_specialized_formula_cache :
+  Rule.taint_rule list ->
+  (Range_with_metavars.ranges * Matching_explanation.t list) option
+  Formula_tbl.t
+
 (* It could be a private function, but it is also used by Deep Semgrep. *)
+(* This [formula_cache] argument is exposed here because this function is also
+   a subroutine but the cache itself should be created outside of the any main
+   loop which runs over rules. This cache is only safe to share with if
+   [taint_config_of_rule] is used on the same file!
+*)
 val taint_config_of_rule :
+  formula_cache:
```

Comment:
It would be nicer having an abstract type alias `formula_cache`, the rest should be an implementation detail.